apisix/plugins/server-info.lua

```diff
@@ -0,0 +1,274 @@
+    end
+
+    server_info.up_time = ngx_time() - boot_time
+    return server_info
+end
+
+
+local function get_server_info()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return 500, err
+    end
+
+    return 200, core.json.encode(server_info)
+end
+
+
+local function report()
+    local server_info, err = get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return
+    end
+
+    if server_info.etcd_version == "unknown" then
+        local res, err = core.etcd.server_version()
+        if not res then
+            core.log.error("failed to fetch etcd version: ", err)
+            return
+
+        elseif type(res.body) ~= "table" then
+            core.log.error("failed to fetch etcd version: bad version info")
+            return
+
+        else
+            server_info.etcd_version = res.body.etcdcluster
+        end
+    end
+
+    server_info.last_report_time = ngx_time()
+
+    local data, err = core.json.encode(server_info)
+    if not data then
+        core.log.error("failed to encode server_info: ", err)
+        return
+    end
+
+    local key = "/data_plane/server_info/" .. server_info.id
+    local ok, err = core.etcd.set(key, data, 180)
+    if not ok then
+        core.log.error("failed to report server info to etcd: ", err)
+        return
+    end
+
+    local ok, err = save(data, false)
+    if not ok then
+        core.log.error("failed to encode and save server info: ", err)
+        return
+    end
+end
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.init()
+    if ngx_worker_id() ~= 0 then
+        -- only let the No.0 worker to launch timer for server info reporting.
+        return
+    end
+
+    local ok, err = encode_and_save(uninitialized_server_info(), true)
+    if not ok then
+        core.log.error("failed to encode and save server info: ", err)
+    end
+
+    if core.config ~= require("apisix.core.config_etcd") then
+        -- we don't need to report server info if etcd is not in use.
+        return
+    end
+
+    local local_conf = core.config.local_conf()
+    local attr = core.table.try_read_attr(local_conf, "plugin_attr",
+                                          plugin_name)
+    local ok, err = core.schema.check(attr_schema, attr)
+    if not ok then
+        core.log.error("failed to check plugin_attr: ", err)
+        return
+    end
+
+    local report_interval = attr.report_interval
+
+    -- we don't use core.timer and the apisix.timers here for:
+    -- 1. core.timer is not cancalable, timers will be leaked if plugin
+    -- reloading happens.
+    -- 2. the background timer in apisix.timers fires per 500 milliseconds, if
+    -- report_interval is not multiple of 0.5, the real report interval will be
```

Comment:
I forgot to remove the second comment.

But the backend timer is not suitable, we use backend timer without caring the fire period, if we use it here, we have to do some calculations according to the 500 milliseconds, which must be updated once the backend timer fire period changed.

Reply:
> we have to do some calculations according to the 500 milliseconds, which must be updated once the backend timer fire period changed

We can expose the `0.5` as a constant, and used in this plugin. It would be better than inventing a new cycle.

Reply:
Well, if we do so, the interval must be a multiple of 0.2, 0.5. I think it should be done in another PR.

Reply:
> Well, if we do so, the interval must be a multiple of 0.2, 0.5. I think it should be done in another PR.

All right, i changed my mind, i will also export the check interval in this PR.